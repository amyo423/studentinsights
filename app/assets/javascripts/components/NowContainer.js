import React from 'react';

// A container for testing that allows passing in a `nowFn`
// that will be shared with child components through context.
export default class NowContainer extends React.Component {
  getChildContext() {
    const {nowFn} = this.props;
    return {
      nowFn() { return nowFn(); }
    };
  }

  render() {
    const {children} = this.props;

    return children;
  }
}
NowContainer.childContextTypes = {
  nowFn: React.PropTypes.func
};
NowContainer.propTypes = {
  children: React.PropTypes.node.isRequired,
  nowFn: React.PropTypes.func.isRequired
};