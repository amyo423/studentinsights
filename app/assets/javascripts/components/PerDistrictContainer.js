import React from 'react';
import PropTypes from 'prop-types';


// A container that provides `districtKey`, assumed to be static across
// the lifetime of all JS execution.  For React components, use this instead of
// `readEnv`, which is intended for bootstrap code.
export default class PerDistrictContainer extends React.Component {
  getChildContext() {
    const {districtKey} = this.props;
    return {districtKey};
  }

  render() {
    const {children} = this.props;
    return children;
  }
}
PerDistrictContainer.childContextTypes = {
  districtKey: PropTypes.string
};
PerDistrictContainer.propTypes = {
  districtKey: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired
};
