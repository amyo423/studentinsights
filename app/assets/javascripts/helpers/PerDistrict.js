import * as Filters from './Filters';


const ORDERED_DISABILITY_VALUES_MAP = {
  'new_bedford': [
    "Does Not Apply",
    "Low-Less Than 2hrs/week",
    "Low-2+ hrs/week",
    "Moderate",
    "High"
  ],
  'somerville': [
    // also include null
    'Low < 2',
    'Low >= 2',
    'Moderate',
    'High'
  ],
  'demo': [
    // also include null
    'Low < 2',
    'Low >= 2',
    'Moderate',
    'High'
  ],

};

export function orderedDisabilityValues(districtKey) {
  return ORDERED_DISABILITY_VALUES_MAP[districtKey] || [];
}

// The table for `SlicePanels` that works differently for different
// districts because the data is stored differently, and this returns
// [{value,filter}] describing which filters the UI should show.
export function slicePanelsDisabilityItems(districtKey) {
  const key = 'sped_level_of_need';
  const itemsFromValues = orderedDisabilityValues(districtKey).map(value => {
<<<<<<< Updated upstream
    return createItemFn(value, Filters.Equal(key, value));
  }, this);
=======
    const filter = Filters.Equal(key, value);
    return {value, filter};
  });
>>>>>>> Stashed changes

  // Somerville uses a null value for no disability, while New Bedford
  // uses a separate value to describe that explicitly.
  return ((districtKey === 'somerville') || (districtKey === 'demo'))
    ? [{value: 'None', filter: Filters.Null(key)}].concat(itemsFromValues)
    : itemsFromValues;
}

// Check educator labels to see if the educator belongs to
// the NGE and 10GE experience teams.  This label only applies to
// Somerville.
export function inExperienceTeam(educatorLabels) {
  return (educatorLabels.indexOf('shs_experience_team') !== -1);
}


const ORDERED_SOMERVILLE_SCHOOL_SLUGS_BY_GRADE = {
  'pic': 10,
  'cap': 100,
  'brn': 200,
  'hea': 300,
  'kdy': 300,
  'afas': 300,
  'escs': 300,
  'wsns': 300,
  'whcs': 300,
  'nw': 400,
  'shs': 500,
  'fc': 500,
  'sped': 600
};

export function sortSchoolSlugsByGrade(districtKey, slugA, slugB) {
  if (districtKey === 'somerville') {
    return ORDERED_SOMERVILLE_SCHOOL_SLUGS_BY_GRADE[slugA] - ORDERED_SOMERVILLE_SCHOOL_SLUGS_BY_GRADE[slugB];
  }

  return slugA.localeCompare(slugB);
}