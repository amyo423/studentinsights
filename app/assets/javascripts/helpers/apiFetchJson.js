// Fetch with common headers
function apiFetch(url, options = {}) {
  const fetchOptions = {
    credentials: 'same-origin',
    headers: {
      'Accept': 'application/json'
    },
    ...options
  };
  return fetch(url, fetchOptions)
    .then(response => response.json());
}


export function apiFetchJson(url) {
  return apiFetch(url, {
    method: 'GET',
    headers: {
      'Accept': 'application/json'
    }
  });
}

// This relies on a Rails CSRF token being rendered on the page
function readCsrfFromPage() {
  return $('meta[name="csrf-token"]').attr('content');
}

// This relies on a Rails CSRF token being rendered on the page
export function apiPostJson(url, body, options = {}) {
  const csrfToken = options.csrfToken || readCsrfFromPage();
  return apiFetch(url, {
    method: 'POST',
    body: JSON.stringify(body),
    headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      'X-CSRF-Token': csrfToken
    },
    ...(options.fetchOptions || {})
  });
}

// This relies on a Rails CSRF token being rendered on the page
export function apiPatchJson(url, body, options = {}) {
  return apiPostJson(url, body, {
    ...options,
    fetchOptions: {
      ...(options.fetchOptions || {}),
      method: 'PATCH'
    }
  });
}

// This relies on a Rails CSRF token being rendered on the page, and
// this adds in query params like Rails' jquery_ujs would.
export function apiDeleteJson(url, options = {}) {
  const csrfToken = options.csrfToken || readCsrfFromPage();
  return fetch(url, {
    method: 'DELETE',
    credentials: 'same-origin',
    body: JSON.stringify({
      _method: 'delete',
      authenticity_token: csrfToken,
      ...(options.body || {})
    }),
    headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      'X-CSRF-Token': csrfToken,
      ...(options.headers || {})
    },
    ...options
  }).then(response => response.json());
}
