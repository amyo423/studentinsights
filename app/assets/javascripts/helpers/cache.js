// as naive as possible
export function createCache() {
  const cacheStorage = {};

  // eg cache('expensiveThing', [this.state, arguments], () => { ... });
  return function cache(keys, fn) {
    const cacheKey = JSON.stringify(keys);
    const cached = cacheStorage[cacheKey];
    if (cached !== undefined) return cached;

    const value = fn();
    cacheStorage[cacheKey] = value;
    return value;
  };
}
