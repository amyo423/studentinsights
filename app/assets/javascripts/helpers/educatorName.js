// Turns SIS format (Watson, Joe) -> Joe Watson and falls back to email if there isn't a full_name
export function formatEducatorName(educator) {
  if (educator.full_name === null) return educator.email.split('@')[0] + '@';
  const parts = educator.full_name.split(', ');
  return parts[1] + ' ' + parts[0];
}