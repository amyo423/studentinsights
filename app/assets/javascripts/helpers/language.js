import _ from 'lodash';
import {SOMERVILLE} from '../helpers/PerDistrict';
import {ALL} from '../components/SimpleFilterSelect';

export const LIMITED_ENGLISH = 'Limited English';
export const FLUENT_ENGLISH = 'Fluent English';

const somervilleMap = {
  'Limited': LIMITED_ENGLISH,
  'Fluent': FLUENT_ENGLISH,
  'FLEP': 'FLEP'
};

const newBedfordMap = {
  'Limited English': LIMITED_ENGLISH,
  'Fluent': FLUENT_ENGLISH,
  'Non-English': 'Non-English',
  'Redesignated': 'Redesignated FLEP',
  'Native': 'Native English'
};

const bedfordMap = {
  'Capable': FLUENT_ENGLISH,
  'Limited English': LIMITED_ENGLISH,
  'Not Capable': LIMITED_ENGLISH
};

export function englishProficiencyOptions(districtKey) {
  if (districtKey === SOMERVILLE) {
    return [
      { value: ALL, label: 'All' },
      { value: 'Fluent', label: FLUENT_ENGLISH },
      { value: 'Limited', label: LIMITED_ENGLISH },
      { value: 'FLEP', label: 'FLEP' }
    ];
  }

  throw new Error(`unsupported districtKey: ${districtKey}`);
}

// This varies by district, but this implementation works across all districts for now.
export function prettyEnglishProficiencyText(limitedEnglishProficiencyValue, access) {
  const prettyTextMap = {
    ...somervilleMap,
    ...newBedfordMap,
    ...bedfordMap
  };
  return (hasAnyAccessData(access) && access.composite)
    ? proficiencyTextForScore(access.composite)
    : prettyTextMap[limitedEnglishProficiencyValue] || 'No LEP status';
}

export function isFluentEnglish(limitedEnglishProficiencyValue) {
  return (prettyEnglishProficiencyText(limitedEnglishProficiencyValue) === FLUENT_ENGLISH);
}

export function hasAnyAccessData(access) {
  return _.some(Object.keys(access), key => access[key]);
}

export function proficiencyTextForScore(compositeDataPoint, options = {}) {
  if (!compositeDataPoint || !compositeDataPoint.performance_level) return 'No data';
  const englishText = options.withoutEnglishText ? '' : ' English';
  const scoreNumber = parseFloat(compositeDataPoint.performance_level);
  if (!scoreNumber) return 'No data';
  if (scoreNumber < 2) return `Entering${englishText}, level 1`;
  if (scoreNumber < 3) return `Emerging${englishText}, level 2`;
  if (scoreNumber < 4) return `Developing${englishText}, level 3`;
  if (scoreNumber < 5) return `Expanding${englishText}, level 4`;
  if (scoreNumber < 6) return `Bridging${englishText}, level 5`;
  if (scoreNumber === 6) return `Reaching${englishText}, level 6`;

  return 'Unknown';
}