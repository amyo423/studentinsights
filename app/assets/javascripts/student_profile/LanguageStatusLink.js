import React from 'react';
import PropTypes from 'prop-types';
import {prettyEnglishProficiencyText, hasAnyAccessData} from '../helpers/language';
import HelpBubble, {
  modalFromLeft,
} from '../components/HelpBubble';
import AccessPanel from './AccessPanel';


// Renders text or link about ELL status, with popup to see more if ELL or FLEP.
// Determine their status solely based on `limited_english_proficiency` field.
// Depending on what that is, there may be additional information from clicking.
export default class LanguageStatusLink extends React.Component {
  render() {
    const {districtKey} = this.context;
    const {limitedEnglishProficiency, access, style} = this.props;
    const el = <div style={style}>{prettyEnglishProficiencyText(districtKey, limitedEnglishProficiency, {access})}</div>;

    // Link to scores also if any access data (even if old)
    if (!hasAnyAccessData(districtKey, limitedEnglishProficiency)) return el;
    return (
      <HelpBubble
        style={{marginLeft: 0, display: 'block'}}
        teaser={el}
        linkStyle={style}
        modalStyle={{
          ...modalFromLeft,
          content: {
            ...modalFromLeft.content,
            width: 750
          }
        }}
        title="Language learning"
        content={<AccessPanel access={access} />}
      />
    );
  }
}
LanguageStatusLink.contextTypes = {
  districtKey: PropTypes.string.isRequired
};
LanguageStatusLink.propTypes = {
  limitedEnglishProficiency: PropTypes.string,
  access: PropTypes.object,
  style: PropTypes.object
};
LanguageStatusLink.defaultProps = {
  style: {}
};
