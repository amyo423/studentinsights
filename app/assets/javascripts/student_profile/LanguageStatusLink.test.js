import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {withDefaultNowContext} from '../testing/NowContainer';
import PerDistrictContainer from '../components/PerDistrictContainer';
import _ from 'lodash';
import {SOMERVILLE, BEDFORD, NEW_BEDFORD} from '../helpers/PerDistrict';
import {TESTING_ONLY_ALL_LANGUAGE_MAPS} from '../helpers/language';
import LanguageStatusLink from './LanguageStatusLink';


export function allCombinations() {
  return _.flatMap([SOMERVILLE, NEW_BEDFORD, BEDFORD], districtKey => {
    const mapForDistrict = TESTING_ONLY_ALL_LANGUAGE_MAPS[districtKey];
    return Object.keys(mapForDistrict).concat([null]).map(limitedEnglishProficiency => {
      return {districtKey, limitedEnglishProficiency};
    });
  });
}

export function testProps(props = {}) {
  return {
    studentFirstName: 'Mari',
    limitedEnglishProficiency: 'Limited',
    access: null,
    style: {
      fontSize: 14
    },
    ...props
  };
}

export function testEl(props, context = {}) {
  const districtKey = context.districtKey || 'somerville';
  return withDefaultNowContext(
    <PerDistrictContainer districtKey={districtKey}>
      <LanguageStatusLink {...props} />
    </PerDistrictContainer>
  );
}

function testRender(props) {
  const el = document.createElement('div');
  ReactDOM.render(testEl(props), el);
  return el;
}

it('renders without crashing', () => {
  testRender(testProps());
});

describe('snapshots', () => {
  const props = testProps();
  const tree = renderer
    .create(testEl(props))
    .toJSON();
  expect(tree).toMatchSnapshot();
});