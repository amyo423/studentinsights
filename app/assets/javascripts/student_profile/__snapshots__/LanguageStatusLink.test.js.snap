// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  style={
    Object {
      "fontSize": 14,
    }
  }
>
  English Learner
</div>
`;
